src/node_perf.h

```diff
@@ -0,0 +1,401 @@
+  V(PRELOAD_MODULE_LOAD_END, "preload_modules_load_end")
+
+enum NodePerformanceEntryTypes {
+#define V(name, _) NODE_PERFORMANCE_ENTRY_TYPE_##name,
+  NODE_PERFORMANCE_ENTRY_TYPES(V)
+#undef V
+  NODE_PERFORMANCE_ENTRY_TYPE_COUNT
+};
+
+enum NodePerformanceMilestones {
+#define V(name, _) NODE_PERFORMANCE_MILESTONE_##name,
+  NODE_PERFORMANCE_MILESTONES(V)
+#undef V
+  // Not an actual milestone.
+  NODE_PERFORMANCE_MILESTONE_COUNT
+};
+
+// The time origin is the starting reference point timestamp
+extern const uint64_t timeOrigin;
+const uint64_t MAX_UINT64 = std::numeric_limits<uint64_t>::max();
+
+NODE_EXTERN inline void PerformanceMilestone(
+    Environment* env,
+    NodePerformanceMilestones milestone) {
+  double* milestones = env->performance_milestones();
+  double now = uv_hrtime();
+  milestones[milestone] = now;
+}
+
+NODE_EXTERN inline void PerformanceIncrementLoop(Environment* env) {
+  double* milestones = env->performance_milestones();
+  if (milestones[NODE_PERFORMANCE_MILESTONE_LOOP_COUNT] < MAX_UINT64)
+    milestones[NODE_PERFORMANCE_MILESTONE_LOOP_COUNT]++;
+}
+
+#define PERFORMANCE_MARK(env, n)                                              \
+  do {                                                                        \
+    PerformanceMilestone(env,                                                 \
+                         node::performance::NODE_PERFORMANCE_MILESTONE_##n);  \
+  } while (0);
+
+class PerformanceEntry : public BaseObject {
+ public:
+  static void New(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void GetName(const v8::Local<v8::String> prop,
+                      const v8::PropertyCallbackInfo<v8::Value>& info);
+  static void GetType(const v8::Local<v8::String> prop,
+                      const v8::PropertyCallbackInfo<v8::Value>& info);
+  static void GetStartTime(const v8::Local<v8::String> prop,
+                           const v8::PropertyCallbackInfo<v8::Value>& info);
+  static void GetDuration(const v8::Local<v8::String> prop,
+                          const v8::PropertyCallbackInfo<v8::Value>& info);
+
+  std::string name() const {
+    return name_;
+  }
+
+  std::string type() const {
+    return type_;
+  }
+
+  virtual double startTime() {
+    return startTime_ / 1e6;
+  }
+
+  virtual double duration() {
+    return durationNano() / 1e6;
+  }
+
+  virtual uint64_t startTimeNano() {
+    return startTime_;
+  }
+
+  virtual uint64_t durationNano() {
+    return refEndTime() - refStartTime();
+  }
+
+  PerformanceEntry* next() {
+    return next_;
+  }
+
+  ~PerformanceEntry() {
+    RemoveFromTimeline(env(), this);
+  }
+
+ protected:
+  PerformanceEntry(Environment* env,
+                   v8::Local<v8::Object> wrap,
+                   const char* name,
+                   const char* type) :
+                   BaseObject(env, wrap),
+                   name_(name),
+                   type_(type),
+                   startTime_(uv_hrtime()) {
+    CallMakeWeak();
```

Comment:
Correct me if I'm wrong, but making the object weak would mean that this object is free to be GC'd when there's no more reference to it that V8 is aware of -- which is the case for `PerformanceEntry`s because the linked list only links the C++ objects together, not using Persistent references, and because the JS side does not kept a reference to the created object.